File: bot/supabase_sync.py
"""
bot/supabase_sync.py
------------------------------------------------------------------------------
Тонкий шар над Supabase REST (PostgREST) для двох таблиць:
    route_alerts  — публічні активні оголошення (їх читає застосунок)
    delay_reports — сирі скарги користувачів (лог подій)

Supabase тут ОСНОВНЕ джерело правди, локальні JSON-файли (data-runtime/,
public/data/route-alerts.json) лишаються як РЕЗЕРВНА копія — пишуться в
telegram_bot.py як і раніше, незалежно від того, чи вдався запит до
Supabase. Якщо SUPABASE_URL/SUPABASE_SERVICE_KEY не задані або якийсь
запит впав — усі функції нижче тихо повертають False/None і бот працює
далі виключно на JSON, як до інтеграції.

Використовується service_role ключ (не anon!) — він обходить RLS, тож
пише в обидві таблиці без обмежень. НІКОЛИ не комітьте цей ключ у git і
не кладіть у фронтенд — тільки в .env бота / секрети Actions.
"""
from __future__ import annotations

import logging
import os
import re
from typing import Any, Optional

import requests

log = logging.getLogger("kharkivgo-bot.supabase")

SUPABASE_URL = os.getenv("SUPABASE_URL", "").strip().rstrip("/")
SUPABASE_SERVICE_KEY = os.getenv("SUPABASE_SERVICE_KEY", "").strip()

ENABLED = bool(SUPABASE_URL and SUPABASE_SERVICE_KEY)

_HEADERS = {
    "apikey": SUPABASE_SERVICE_KEY,
    "Authorization": f"Bearer {SUPABASE_SERVICE_KEY}",
    "Content-Type": "application/json",
}

if not ENABLED:
    log.info("SUPABASE_URL/SUPABASE_SERVICE_KEY не задано — синхронізація вимкнена, працюю тільки з JSON.")
elif re.search(r"/(rest|auth|storage)(/|$)", SUPABASE_URL, re.IGNORECASE):
    log.warning(
        "SUPABASE_URL схоже містить зайвий шлях (%s) — має бути лише базовий домен проєкту, "
        "напр. https://xxxx.supabase.co, без /rest/v1 чи іншого хвоста.",
        SUPABASE_URL,
    )


def _rest(method: str, table: str, *, params: Optional[dict] = None, json_body: Any = None, prefer: str = "") -> Optional[Any]:
    if not ENABLED:
        return None
    headers = dict(_HEADERS)
    if prefer:
        headers["Prefer"] = prefer
    try:
        res = requests.request(
            method,
            f"{SUPABASE_URL}/rest/v1/{table}",
            headers=headers,
            params=params,
            json=json_body,
            timeout=15,
        )
        if res.status_code >= 400:
            log.warning("Supabase %s %s -> %s: %s", method, res.url, res.status_code, res.text[:300])
            return None
        if not res.text:
            return []
        return res.json()
    except requests.RequestException as e:
        log.warning("Supabase %s %s мережева помилка: %s", method, table, e)
        return None


def replace_route_alerts(alerts: list[dict]) -> bool:
    """Повністю синхронізує таблицю route_alerts зі списком активних
    оголошень (той самий список, що йде у route-alerts.json). Проста
    стратегія "видалити все й вставити поточне" — таблиця мала (десятки
    рядків максимум), тож дешевше й надійніше за диффінг по id."""
    if not ENABLED:
        return False

    # Видаляємо все (id завжди > 0, unix ms timestamp)
    if _rest("DELETE", "route_alerts", params={"id": "gt.0"}) is None:
        return False

    if not alerts:
        return True

    rows = [
        {
            "id": a["id"],
            "kind": a.get("kind"),
            "route_number": a["routeNumber"],
            "message": a["message"],
            "created_at": a["createdAt"],
            "expires_at": a["expiresAt"],
            "source": a.get("source", "manual"),
        }
        for a in alerts
    ]
    result = _rest("POST", "route_alerts", json_body=rows, prefer="resolution=merge-duplicates,return=minimal")
    return result is not None


def insert_delay_reports(new_reports: list[dict]) -> bool:
    """Додає в лог тільки НОВІ скарги цього циклу (delay_reports —
    append-only лог подій, весь список переписувати не треба)."""
    if not ENABLED or not new_reports:
        return False

    rows = [
        {
            "user_id": r.get("userId"),
            "username": r.get("username"),
            "kind": r.get("kind"),
            "route_number": r["routeNumber"],
            "comment": r.get("comment") or None,
            "created_at": r["createdAt"],
        }
        for r in new_reports
    ]
    result = _rest("POST", "delay_reports", json_body=rows, prefer="return=minimal")
    return result is not None

File: frontend/scripts/supabaseSync.mjs
/**
 * scripts/supabaseSync.mjs
 * ---------------------------------------------------------------------------
 * Те саме, що bot/supabase_sync.py, але для GitHub Actions-версії бота
 * (process-telegram-bot.mjs). Supabase — основне джерело правди для
 * route_alerts/delay_reports, JSON-файли лишаються резервною копією і
 * пишуться як і раніше незалежно від результату синхронізації.
 *
 * SUPABASE_URL / SUPABASE_SERVICE_KEY беруться зі секретів репозиторію
 * (Settings → Secrets and variables → Actions). Якщо не задані — усі
 * функції нижче тихо повертають false, бот працює тільки на JSON.
 */

const SUPABASE_URL = (process.env.SUPABASE_URL || '').replace(/\/$/, '');
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_KEY || '';

export const SUPABASE_ENABLED = Boolean(SUPABASE_URL && SUPABASE_SERVICE_KEY);

if (!SUPABASE_ENABLED) {
  console.log('[bot] SUPABASE_URL/SUPABASE_SERVICE_KEY не задано — синхронізація вимкнена, працюю тільки з JSON.');
}

async function rest(method, table, { params, body, prefer } = {}) {
  if (!SUPABASE_ENABLED) return null;
  const url = new URL(`${SUPABASE_URL}/rest/v1/${table}`);
  if (params) {
    for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  }
  const headers = {
    apikey: SUPABASE_SERVICE_KEY,
    Authorization: `Bearer ${SUPABASE_SERVICE_KEY}`,
    'Content-Type': 'application/json'
  };
  if (prefer) headers.Prefer = prefer;

  try {
    const res = await fetch(url, {
      method,
      headers,
      body: body !== undefined ? JSON.stringify(body) : undefined
    });
    if (!res.ok) {
      console.warn(`[bot] Supabase ${method} ${table} -> ${res.status}: ${(await res.text()).slice(0, 300)}`);
      return null;
    }
    const text = await res.text();
    return text ? JSON.parse(text) : [];
  } catch (err) {
    console.warn(`[bot] Supabase ${method} ${table} мережева помилка:`, err?.message || err);
    return null;
  }
}

/** Повністю замінює вміст route_alerts поточним списком активних оголошень. */
export async function replaceRouteAlerts(alerts) {
  if (!SUPABASE_ENABLED) return false;

  const deleted = await rest('DELETE', 'route_alerts', { params: { id: 'gt.0' } });
  if (deleted === null) return false;

  if (!alerts.length) return true;

  const rows = alerts.map((a) => ({
    id: a.id,
    kind: a.kind,
    route_number: a.routeNumber,
    message: a.message,
    created_at: a.createdAt,
    expires_at: a.expiresAt,
    source: a.source || 'manual'
  }));
  const result = await rest('POST', 'route_alerts', { body: rows, prefer: 'resolution=merge-duplicates,return=minimal' });
  return result !== null;
}

/** Додає в лог лише нові скарги цього циклу (append-only). */
export async function insertDelayReports(newReports) {
  if (!SUPABASE_ENABLED || !newReports.length) return false;

  const rows = newReports.map((r) => ({
    user_id: r.userId,
    username: r.username || null,
    kind: r.kind,
    route_number: r.routeNumber,
    comment: r.comment || null,
    created_at: r.createdAt
  }));
  const result = await rest('POST', 'delay_reports', { body: rows, prefer: 'return=minimal' });
  return result !== null;
}
